arbnode/blockmetadata.go

```diff
@@ -0,0 +1,149 @@
+	f.Duration(prefix+".sync-interval", DefaultBlockMetadataFetcherConfig.SyncInterval, "interval at which blockMetadata are synced regularly")
+	f.Uint64(prefix+".api-blocks-limit", DefaultBlockMetadataFetcherConfig.APIBlocksLimit, "maximum number of blocks allowed to be queried for blockMetadata per arb_getRawBlockMetadata query.\n"+
+		"This should be set lesser than or equal to the limit on the api provider side")
+}
+
+type BlockMetadataFetcher struct {
+	stopwaiter.StopWaiter
+	config BlockMetadataFetcherConfig
+	db     ethdb.Database
+	client *rpcclient.RpcClient
+	exec   execution.ExecutionClient
+}
+
+func NewBlockMetadataFetcher(ctx context.Context, c BlockMetadataFetcherConfig, db ethdb.Database, exec execution.ExecutionClient) (*BlockMetadataFetcher, error) {
+	client := rpcclient.NewRpcClient(func() *rpcclient.ClientConfig { return &c.Source }, nil)
+	if err := client.Start(ctx); err != nil {
+		return nil, err
+	}
+	return &BlockMetadataFetcher{
+		config: c,
+		db:     db,
+		client: client,
+		exec:   exec,
+	}, nil
+}
+
+func (b *BlockMetadataFetcher) fetch(ctx context.Context, fromBlock, toBlock uint64) ([]gethexec.NumberAndBlockMetadata, error) {
+	var result []gethexec.NumberAndBlockMetadata
+	err := b.client.CallContext(ctx, &result, "arb_getRawBlockMetadata", rpc.BlockNumber(fromBlock), rpc.BlockNumber(toBlock))
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+func (b *BlockMetadataFetcher) persistBlockMetadata(query []uint64, result []gethexec.NumberAndBlockMetadata) error {
+	batch := b.db.NewBatch()
+	queryMap := util.ArrayToMap(query)
+	for _, elem := range result {
+		pos, err := b.exec.BlockNumberToMessageIndex(elem.BlockNumber)
+		if err != nil {
+			return err
+		}
+		if _, ok := queryMap[uint64(pos)]; ok {
+			if err := batch.Put(dbKey(blockMetadataInputFeedPrefix, uint64(pos)), elem.RawMetadata); err != nil {
+				return err
+			}
+			if err := batch.Delete(dbKey(missingBlockMetadataInputFeedPrefix, uint64(pos))); err != nil {
+				return err
+			}
+			// If we reached the ideal batch size, commit and reset
+			if batch.ValueSize() >= ethdb.IdealBatchSize {
+				if err := batch.Write(); err != nil {
+					return err
+				}
+				batch.Reset()
+			}
+		}
+	}
+	return batch.Write()
+}
+
+func (b *BlockMetadataFetcher) Update(ctx context.Context) time.Duration {
+	handleQuery := func(query []uint64) bool {
+		result, err := b.fetch(
+			ctx,
+			b.exec.MessageIndexToBlockNumber(arbutil.MessageIndex(query[0])),
+			b.exec.MessageIndexToBlockNumber(arbutil.MessageIndex(query[len(query)-1])),
+		)
+		if err != nil {
+			log.Error("Error getting result from bulk blockMetadata API", "err", err)
+			return false
+		}
+		if err = b.persistBlockMetadata(query, result); err != nil {
+			log.Error("Error committing result from bulk blockMetadata API to ArbDB", "err", err)
+			return false
+		}
+		return true
+	}
+	iter := b.db.NewIterator(missingBlockMetadataInputFeedPrefix, nil)
```

Comment:
I think it's worth pointing out that in the event of missing block metadata that is not present in the `Source` this will retry fetching that data every 5 minutes. I don't think it's a problem but I want to flag this behavior.

Reply:
added this